File: src/main/java/com/github/te170476/musicdsl/score/note/NoteValue.java
package com.github.te170476.musicdsl.score.note;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class NoteValue {
    public static INoteValue get(int... index) {
        return new Values(
                Arrays.stream(index)
                        .mapToObj(Value::get)
                        .collect(Collectors.toList())
        );
    }
    public static INoteValue get(INoteValue... index) {
        return new Values(
                Arrays.stream(index)
                        .collect(Collectors.toList())
        );
    }

    public interface LeafNode extends INoteValue {}
    public interface InnerNode extends INoteValue {}

    static class Value implements LeafNode {
        public final int value;

        public static Value get(int index) {
            return new Value(index);
        }
        private Value(int noteValue) {
            this.value = noteValue;
        }

        @Override
        public double toPercentage() {
            return value == 0 ? 0.0 : 1.0 / value;
        }
    }
    
    static class Values implements InnerNode {
        public List<INoteValue> leaves;
        public Values(List<INoteValue> leaves) {
            this.leaves = leaves;
        }

        @Override
        public double toPercentage() {
            return leaves.stream()
                    .mapToDouble(INoteValue::toPercentage)
                    .sum();
        }
    }
    static class HasBase implements InnerNode {
        public INoteValue base;
        public INoteValue value;
        public HasBase(INoteValue base, INoteValue value) {
            this.base = base;
            this.value = value;
        }

        @Override
        public double toPercentage() {
            return value.toPercentage() * base.toPercentage();
        }
    }
    static class Negative implements InnerNode {
        public INoteValue value;
        public Negative(INoteValue value) {
            this.value = value;
        }
        @Override
        public double toPercentage() {
            return value.toPercentage() * -1;
        }
    }
}

File: src/main/scala/com/github/nnnnusui/musicdsl/entity/score/note/NoteValue.scala
package com.github.nnnnusui.musicdsl.entity
package score.note

object NoteValue {
  private val _0: NoteValue = Value(0)
  def apply(value: Int*): NoteValue =
    value
      .map(it => Value(it))
      .foldLeft(_0) { (it, sum) => sum + it }

  private case class Value(value: Int) extends NoteValue {
    override def percentage: Double =
      value match {
        case 0  => 0.0
        case it => 1.0 / it
      }
  }
  private case class Values(values: Seq[NoteValue]) extends NoteValue {
    override def percentage: Double = values.map(_.percentage).sum
  }
  private case class HasBase(base: NoteValue, value: NoteValue) extends NoteValue {
    override def percentage: Double = value.percentage * base.percentage
  }
}
sealed trait NoteValue {
  def percentage: Double
  private def base(base: NoteValue): NoteValue = NoteValue.HasBase(base, this)

  def +(rhs: NoteValue): NoteValue = NoteValue.Values(Seq(this, rhs))
  def +(rhs: Int): NoteValue = this + NoteValue(rhs)
  def *(rhs: Int): NoteValue = NoteValue.Values(Seq.fill(rhs)(this))
  def /(rhs: NoteValue): NoteValue = this.base(rhs)
  def /(rhs: Int): NoteValue = this / NoteValue(rhs)
}
